### PR TITLE
Add jest and unit tests to everything

### PR DESCRIPTION
**Pull request by knewkirk**
# Add some unit testing _(finally)_
- also includes minor nits I found while testing that don't impact functionality
- will run tests on pre-commit

Screen Shot 2022-12-21 at 4 20 55 PM

### CODE REVIEW

File: jest.config.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  // forceCoverageMatch: [],
+
+  // A path to a module which exports an async function that is triggered once before all test suites
+  // globalSetup: undefined,
+
+  // A path to a module which exports an async function that is triggered once after all test suites
+  // globalTeardown: undefined,
+
+  // A set of global variables that need to be available in all test environments
+  globals: { wally: 'whatever' },
```
Comment (wchan2) [inline jest.config.ts:68]: Q: Is this needed?

File: test/auth.test.ts [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+      });
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+    });
+
+    it('rejects if already logging in', () => {
+      auth.isLoggingIn = true;
+      return expect(auth.login()).rejects.toBeDefined();
```
Comment (wchan2) [inline test/auth.test.ts:67]: This test seems to indicate that it returns just a promise or a type that has the `reject` method. I think there's something like `await expect(auth.login()).rejects.ToThrow()` that expects that the promise is actually rejecting.

File: test/auth.test.ts [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+      const consoleSpy = jest.spyOn(console, 'error');
+      consoleSpy.mockImplementation(() => {});
+      const codeGenSpy = jest.spyOn(Auth.prototype as any, 'generateStateCode');
+      auth = new Auth({} as any);
+      auth.login();
+      expect(consoleSpy).toHaveBeenCalled();
+      expect(codeGenSpy).toHaveBeenCalledTimes(0);
+
+      consoleSpy.mockRestore();
+      codeGenSpy.mockRestore();
```
Comment (wchan2) [inline test/auth.test.ts:80]: Minor and non-blocking: Similar comment as the above. Could improve the readability a bit to move the spy setup and teardown into the `beforeEach` and `afterEach`.

File: test/auth.test.ts [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+      expect(windowOpenSpy).toHaveBeenCalledWith(
+        `${host}/oauth/otp?clientId=${clientId}&state=${state}&redirectUrl=${encodeURIComponent(
+          redirectURL
+        )}&email=${encodeURIComponent(email)}`,
+        '_blank'
+      );
+      genCodeSpy.mockRestore();
+    });
+
+    describe('returned promise', () => {
```
Comment (wchan2) [inline test/auth.test.ts:130]: Q: Is it possible to test that the events are triggered? It'll make the test stronger. At the same time, it doesn't have to test internal implementation details as long as the callback is triggered after some action.

File: test/auth.test.ts [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+        expect(workerSpy.mock.calls[0][0]).toBe(WorkerMessage.LOGIN_SUCCESS);
+        expect(workerSpy.mock.calls[1][0]).toBe(WorkerMessage.LOGIN_FAILURE);
+
+        emitterSpy.mockRestore();
+        workerSpy.mockRestore();
+      });
+
+      it('resolves on the accounts changed emitter message', async () => {
+        setTimeout(() => {
+          auth['messenger'].emit(EmitterMessage.ACCOUNTS_CHANGED, '0x1234');
```
Comment (wchan2) [inline test/auth.test.ts:146]:
```suggestion
          auth.messenger.emit(EmitterMessage.ACCOUNTS_CHANGED, '0x1234');
```

File: test/auth.test.ts [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        }, 0);
+
+        await auth.login();
```
Comment (wchan2) [inline test/auth.test.ts:149]: Q: Is there an assertion missing here?

File: test/auth.test.ts [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+
+      it('resolves on the login success worker message when token', async () => {
+        const tokenSpy = jest.spyOn(Auth.prototype, 'getToken');
+        tokenSpy.mockImplementation(() => 'asdf');
+
+        setTimeout(() => {
+          auth['messenger'].handleWorkerMessage(WorkerMessage.LOGIN_SUCCESS);
+        }, 0);
+
+        await auth.login();
```
Comment (wchan2) [inline test/auth.test.ts:160]: Q: Is the token returned here? If so, since `getToken` is mocked, a stronger test would just be checking that the return value is the mocked return value. Generally, testing for `toHaveBeenCalled` requires implementation details and can produce more fragile tests.

File: test/auth.test.ts [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+      auth = undefined as any;
+      storageGetSpy.mockRestore();
+      storageSetSpy.mockRestore();
+      storageRemoveSpy.mockRestore();
+    });
+
+    it.each([
+      ['whatever', true],
+      [null, false],
+    ])('calls getState to determine if redirected', (state, expected) => {
```
Comment (wchan2) [inline test/auth.test.ts:228]: Minor: The curried call can also take a formatted string that also outputs the input values. It makes it easier to distinguish which case may be failing.

File: test/auth.test.ts [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+        consoleSpy = jest.spyOn(console, 'error');
+        consoleSpy.mockImplementation(() => {});
+      });
+
+      afterEach(() => {
+        consoleSpy.mockRestore();
+      });
+
+      it('shows error, deletes state, and sends message on bad response', async () => {
+        const messengerSpy = jest.spyOn(auth['messenger'], 'sendWorkerMessage');
```
Comment (wchan2) [inline test/auth.test.ts:294]:
```suggestion
        const messengerSpy = jest.spyOn(auth.messenger, 'sendWorkerMessage');
```

File: test/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        .spyOn(WallyJS.prototype, 'isRedirected')
+        .mockImplementation(() => true);
+      api.init({ clientId: '1234' });
+      const mockWally = (WallyJS as jest.Mock).mock.instances[0];
+      expect(mockWally.handleRedirect).toHaveBeenCalled();
+    });
+  });
+
+  describe('getProvider', () => {
+    it('returns undefined and shows error when no module instance', () => {
```
Comment (wchan2) [inline test/index.test.ts:64]: Minor: Perhaps it may be better to return `null`. `null` represents an `empty` value whereas `undefined` is more of a possible error case that is not handled.

File: test/messenger.test.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        Messenger.prototype,
+        'handleWorkerMessage'
+      );
+      handleWorkerSpy.mockImplementation(jest.fn());
+      const messenger = new Messenger({
+        sharedWorkerUrl,
+      });
+      expect(SharedWorker).toHaveBeenCalledWith(sharedWorkerUrl);
+      expect(mockStart).toHaveBeenCalled();
+      (messenger['worker'] as any)['port']['onmessage']({ data: 'foo' });
```
Comment (wchan2) [inline test/messenger.test.ts:28]: Q: How come this requires a cast to any? Can we just do the following here?

```suggestion
      messenger.worker.port.onmessage({ data: 'foo' });
```

File: test/messenger.test.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    const name1 = 'connect';
+    const name2 = 'disconnect';
+    const fn1 = (a: any) => {};
+    const fn2 = (b: any) => {};
+
+    beforeEach(() => {
+      messenger = new Messenger({});
+    });
+
+    it('can add listeners', () => {
```
Comment (wchan2) [inline test/messenger.test.ts:61]: These tests seem to be testing the internal implementation heavily. It'll be better to test the actual triggers and then once removed, the functions are no longer triggered.

File: test/messenger.test.ts [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+      }
+    );
+  });
+
+  describe('message emitter', () => {
+    let messenger: Messenger;
+    const name1 = 'connect';
+    const name2 = 'disconnect';
+    const fn1 = (a: any) => {};
+    const fn2 = (b: any) => {};
```
Comment (wchan2) [inline test/messenger.test.ts:55]: Non-blocking: If testing for functionality, `jest.fn()` offers a little bit more :).

File: .eslintrc.js [hunk 14]
```diff
@@ -14,5 +14,5 @@ module.exports = {
     es2021: true,
     node: true,
   },
-  ignorePatterns: ['dist/', 'test_page/'],
+  ignorePatterns: ['dist/', 'test_page/', 'test/'],
```
[the remark this answers is not in the document]
Comment (knewkirk) [reply .eslintrc.js:17]: ah good call

File: jest.config.ts (hunk 1 above)
Comment (knewkirk) [reply jest.config.ts:68]: nope! 😅

File: test/auth.test.ts [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+      new Auth({
+        clientId,
+        host: 'http://localhost:1738',
+        authToken,
+        messenger: new Messenger({}),
+      });
+      expect(storageSpy).toHaveBeenCalledWith('wally:1234:token', authToken);
+
+      storageSpy.mockRestore();
```
[the remark this answers is not in the document]
Comment (knewkirk) [reply test/auth.test.ts:41]: for sure. is that generally the play - put stuff in `beforeEach`/`afterEach` whenever possible, even for a single test?

[the remark this answers is not in the document]
Comment (wchan2) [reply test/auth.test.ts:41]: I generally do that because it separates it and makes it more clear but do also understand that its somewhat subjective so I would say, it's a non-blocking comment for me :).

File: test/auth.test.ts (hunk 9 above)
Comment (knewkirk) [reply test/auth.test.ts:294]: ah yeah - it's to access private members without typescript complaining 🤷

File: test/messenger.test.ts (hunk 11 above)
Comment (knewkirk) [reply test/messenger.test.ts:28]: same re: typescript being needy 😕

File: test/messenger.test.ts (hunk 12 above)
Comment (knewkirk) [reply test/messenger.test.ts:61]: haha yeah fair enough 😄

File: test/auth.test.ts (hunk 5 above)
Comment (knewkirk) [reply test/auth.test.ts:146]: same re: typescript

File: test/auth.test.ts (hunk 6 above)
Comment (knewkirk) [reply test/auth.test.ts:149]: well, the promise returned by `login()` doesn't resolve with any value, so the internet said a simple `await` was basically the same as asserting a resolution. But I agree, it definitely looks a little funky, i'll make it more explicit

File: test/auth.test.ts (hunk 7 above)
Comment (knewkirk) [reply test/auth.test.ts:160]: `login()` returns a `Promise<void>` - I'm not sure if returning the token would be a good/bad idea? Should the consumer of the API get access to it, or is it better kept internal?

otherwise, similar answer to above, i'm mainly testing that the promise is resolved - i can remove the `toHaveBeenCalled`, not super necessary to the test

File: test/auth.test.ts (hunk 8 above)
Comment (knewkirk) [reply test/auth.test.ts:228]: great call 💡

File: test/auth.test.ts (hunk 4 above)
Comment (knewkirk) [reply test/auth.test.ts:130]: Alright, I made one test for the happy path all the way through the login flow, hopefully that's what you're talking about. I know testing implementation details isn't awesome 😅  but I'd like to leave those tests in to explicitly handle edge cases/error handling if that's alright

File: test/messenger.test.ts (hunk 12 above)
Comment (knewkirk) [reply test/messenger.test.ts:61]: i think i'm gonna skip doing the same with the `SharedWorker` though, since it's not defined in the test environment and we'd have to stub implementation details

**Comment by knewkirk** [comment]
Thanks again for the review @wchan2! 

Regarding testing implementation details - I know, I'm honestly pretty terrible at writing tests _(as I'm sure you can tell)_. In my experience, unit tests have just been a way for me to double check the implementation; they end up getting changed by folks whenever they update implementation details anyways. 

For integration tests, I 100% agree with you - you shouldn't have to know implementation details and testing the intention is much more important - maybe y'all can add those later 😄 

I did add a couple higher-level tests like you suggested, hopefully that's ok for now.

_Other highlights:_
- better beforeEach/afterEach consolidation
- better it.each testname formatting
- clearer promise resolution testing
- change getProvider to return null
- better end-to-end message testing
- better end-to-end login testing
- remove old/unnecessary config stuff

File: test/requester.test.ts [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+          },
+        },
+      ],
+    ])('logs an error and rejects when response is %o', (resp) => {
+      global.fetch = jest.fn(() => Promise.resolve(resp) as any);
+      expect.assertions(1);
+      return requester.request({ method: RPCMethodName.BLOCK_NUMBER }).then(
+        () => {
+          expect(true).toBe(false);
+        },
```
Comment (wchan2) [inline test/requester.test.ts:378]: Q: Is this supposed to be called or are you using this such that if it calls into this, it'll error out? I think the `expect.assertions(1)` should also cover that case no?

**Review by wchan2 — APPROVED** [review]
Generally, the tests seem fine. One more minor comment. I would advocate for not reaching in and directly testing `private` methods since there's probably a reason why they're private. I'll leave it up to you and the team to decide what's more important though 😄 .

File: test/requester.test.ts (hunk 16 above)
Comment (knewkirk) [reply test/requester.test.ts:378]: haha - that's a very good point 😆
